syntax = "proto3";

// A message from the browser to the server.
message BackMsg {
  oneof type {
    // Set to true to ask the server to display the inline help.
    // bool help = 1;  DEPRECATED

    // Set to true to ask the server to upload the report.
    bool cloud_upload = 2;

    // Asks the server to run the script with this given command line.
    string rerun_script = 3;

    // DEPRECATED - Asks the server to run the script with this object
    // ReRun rerun = 4;

    // Requests that the report's @st_cache be cleared
    bool clear_cache = 5;

    // Requests that the runOnSave behavior for this report be set
    // to the given value
    bool set_run_on_save = 6;

    // Requests that the report's execution be stopped
    bool stop_report = 7;

    // string widget_json = 8;

    WidgetStates update_widgets = 9;

    // Set to true to ask the server to close the connection
    bool close_connection = 10;
  }
}

// Contains updated state for each widget in a report
message WidgetStates {
  repeated WidgetState widgets = 1;
}

// State for a single widget
message WidgetState {
  string id = 1;

  oneof value {
    // trigger_value is for buttons. A button's value needs to
    // auto-revert back to False after the report has been run with
    // the value set to True. After running the report, the server
    // will reset all trigger_values in its stored Widget state dict.
    // (Use bool_value for widgets like checkboxes, whose state persists
    // beyond a single report run.)
    bool trigger_value = 2;

    bool bool_value = 3;
    float float_value = 4;
    int32 int_value = 5;
    string string_value = 6;
    FloatArray float_array_value = 7;
  }
}

message FloatArray {
  repeated float value = 1;
}
